[MKLDNN] Fuse Conv + Bias using Pass

Sand3r- opened the pull request:
The `test_analyzer_resnet50` was failing because MKLDNN wasn't enabled for it.

luotao1 commented:
> The test_analyzer_resnet50 was failing because MKLDNN wasn't enabled for it.

Though you add `config._use_mkldnn=True` to pass the unittest on CI. But when `config._use_mkldnn=false`, i.e, Analysis Mode in MKL-ONLY environment, there is accuracy diff.

Maybe you should adjust the `graph_pattern_detector.cc`, since this pattern could be detected in MKL-ONLY environment as well.

Sand3r- commented:
> Maybe you should adjust the graph_pattern_detector.cc, since this pattern could be detected in MKL-ONLY environment as well.

@luotao1 Whether this is an MKL-ONLY environment cannot be detected during the stage of passes, since switching the MKL-DNN operator's on, by enabling their `use_mkldnn` attribute happens later, after the passes are carried out.

One way to solve this would be to make running passes and MKL-DNN operators dependent on an environmental variable, such as a `FLAGS_use_mkldnn=1` flag we used to use in python. That way I could detect that MKL-DNN passes should or shouldn't be applied, and whether the operators should be MKL-DNNish or not.

luotao1 commented on paddle/fluid/framework/ir/graph_pattern_detector.cc:
```diff
@@ -964,6 +964,39 @@ PDNode *patterns::ElewiseAddActInplaceGrad::operator()(
   return ele_add_grad;
 }
 
+PDNode *patterns::ConvBias::operator()(
```
Could you detect graph_pattern with the op's attribute `use_mkldnn=true`?

Sand3r- replied:
As I've mentioned - setting of this parameter occurs after the passes are applied. In other words - it will always be false at the stage of using passes.

luotao1 replied:
Got it. Maybe we should implement MKLDNN OP Placement PASS at first to avoid this problem.

Sand3r- replied:
Placement Pass would certainly be a solution to the problem. Perhaps for now we could run 2 tests as it used to be done in older version of `test_analyzer_ocr`:
- config._use_mkldnn = true with mkldnn passes enabled
- config._use_mkldnn = false with mkldnn passes disabled (added to exception list)

luotao1 replied:
I add #13909 to run 2 tests.

Sand3r- replied:
Good job! If it gets accepted, I'll just remove non-mkldnn passes at 
https://github.com/PaddlePaddle/Paddle/pull/13909/files#diff-2e425e802cdeb2a590bc1e5ac9e34f11R23 
and 
https://github.com/PaddlePaddle/Paddle/pull/13909/files#diff-2e425e802cdeb2a590bc1e5ac9e34f11R23 
if the `_use_mkldnn` parameter will be set to true.

luotao1 replied:
You should not remove non-mkldnn passes in the unit-tests, but should update the codes to pass both mkldnn/non-mkldnn passes in the unit-tests.

Sand3r- replied:
How can I detect in graph detector whether the MKL-DNN passes are to be used or not when by default all `use_mkldnn` attributes are set to false?

luotao1 replied:
@Superjomn @tensor-tang Do you have some suggestions about it?

sfraczek replied:
@wojtuss has created a pass that sets mkldnn attribute to operators. He will create a pull request today. If it is accepted, @Sand3r- you could use it.

Sand3r- replied:
@luotao1 Since this is the blocker for the #13485 could we possibly follow this scenario?:
1. Once your #13909 will get merged, I'll apply the temporarily fix for the solution by disabling the MKL-DNN Passes when the `_use_mkldnn` flag is off, by adding them to exception list in config.
2. When @wojtuss will prepare his PR with pass setting use_mkldnn flag to true, I'll remove these pass exceptions from tests and adjust the code accordingly to your will - that is - the value of use_mkldnn parameter will be checked at the stage of detecting patterns in `graph_pattern_detector.cc`.

luotao1 replied:
@Sand3r- Since @wojtuss will create a pull request today, could you wait a few more hours?

Sand3r- replied:
As long as that solution is satisfactory for you, I see no trouble :wink:

luotao1 commented:
@Sand3r- #13909  and #13949 are merged already. Please solve the conflict and pass the unit-tests, thanks!
Besides, could you fix https://github.com/PaddlePaddle/Paddle/pull/13958#pullrequestreview-166367767 in your PR?

Sand3r- commented:
@luotao1 I have rebased, resolved conflicts and adjusted the code so that it correctly builds and runs and passes all analyzer tests as well as conv_relu test successfully on my local machine.

I haven't updated the https://github.com/PaddlePaddle/Paddle/pull/13958#pullrequestreview-166367767 since @wojtuss  has already took care of it in https://github.com/PaddlePaddle/Paddle/pull/13992

Updated: I have ran all of the tests locally and all passed (apart from those that typically fail in debug mode, which are completely unrelated to these code changes).

kbinias commented:
luotao could you possibly continue you review or merge this PR ?

luotao1 commented on paddle/fluid/inference/tests/api/analyzer_resnet50_tester.cc:
```diff
@@ -27,6 +27,9 @@ void SetConfig(AnalysisConfig *cfg) {
   cfg->specify_input_name = true;
+#ifdef PADDLE_WITH_MKLDNN
+  cfg->_use_mkldnn = true;
+#endif
```
line30-32 are not worked, since `cfg._use_mkldnn` would be reset in
https://github.com/PaddlePaddle/Paddle/blob/a1d3db031beff0a47278991d18e7015590ffb1dd/paddle/fluid/inference/tests/api/analyzer_resnet50_tester.cc#L91-L94
You can remove these lines.
